modulemd/v2/modulemd-translation-entry.c

```diff
@@ -0,0 +1,623 @@
+    {
+      YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+
+      switch (event.type)
+        {
+        case YAML_MAPPING_START_EVENT: in_map = TRUE; break;
+
+        case YAML_MAPPING_END_EVENT:
+          in_map = FALSE;
+          done = TRUE;
+          break;
+
+        case YAML_SCALAR_EVENT:
+          if (!in_map)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Missing mapping in translation entry profiles");
+              break;
+            }
+
+          profile_name = g_strdup ((const gchar *)event.data.scalar.value);
+          value = modulemd_yaml_parse_string (parser, error);
+          g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+          modulemd_translation_entry_set_profile_description (
+            te, profile_name, value);
+          g_free (profile_name);
+          g_free (value);
+          g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+          break;
+
+        default:
+          /* We received a YAML event we shouldn't expect at this level */
+          MMD_YAML_ERROR_EVENT_EXIT (
+            error,
+            event,
+            "Unexpected YAML event in translation entry profile");
+          break;
+        }
+      yaml_event_delete (&event);
+    }
+  return NULL;
+}
+
+ModulemdTranslationEntry *
+modulemd_translation_entry_parse_yaml (yaml_parser_t *parser, GError **error)
+{
+  MODULEMD_INIT_TRACE ();
+  MMD_INIT_YAML_EVENT (event);
+  gboolean done = FALSE;
+  gboolean in_map = FALSE;
+  gchar *value;
+  g_autoptr (ModulemdTranslationEntry) te = NULL;
+
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  /* Read in the locale of the translation entry */
+  YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+  if (event.type != YAML_SCALAR_EVENT)
+    {
+      MMD_YAML_ERROR_EVENT_EXIT (
+        error, event, "Missing translation entry locale");
+    }
+  te = modulemd_translation_entry_new ((const gchar *)event.data.scalar.value);
+  yaml_event_delete (&event);
+
+  /* Read in any supplementary attributes of the translation entry
+   */
+  while (!done)
+    {
+      YAML_PARSER_PARSE_WITH_EXIT (parser, &event, error);
+
+      switch (event.type)
+        {
+        case YAML_MAPPING_START_EVENT: in_map = TRUE; break;
+
+        case YAML_MAPPING_END_EVENT:
+          in_map = FALSE;
+          done = TRUE;
+          break;
+
+        case YAML_SCALAR_EVENT:
+          if (!in_map)
+            {
+              MMD_YAML_ERROR_EVENT_EXIT (
+                error, event, "Missing mapping in translation entry");
+              break;
+            }
+          if (!g_strcmp0 ((const gchar *)event.data.scalar.value, "summary"))
+            {
+              value = modulemd_yaml_parse_string (parser, error);
+              modulemd_translation_entry_set_summary (te, value);
+              g_free (value);
+              g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
```

Comment:
Do not use this for logic issues. This is a detection of programming errors. (It reads "the error is either NULL (caller doesn't care about it) or it's non-NULL when initially passed into this function, meaning an error occurred earlier and wasn't handled then".)

See `ServiceLevel.emit_yaml()`. It would be better to create a new nested_error and then pass that to 

You should instead be checking that `value` is non-null and calling ` g_propagate_prefixed_error()` to set `error`.

Then just call `return NULL` here. The auto-pointer cleanups should handle everything else.

(I do realize I forgot to go back and do this in ServiceLevel.parse_yaml(), which I will fix.)